src/QirRuntime/lib/Tracer/tracer.hpp

```diff
@@ -0,0 +1,214 @@
+        }
+        bool AreEqualResults(Result r1, Result r2) override
+        {
+            throw std::logic_error("Cannot compare results while tracing!");
+        }
+        ResultValue GetResultValue(Result result) override
+        {
+            throw std::logic_error("Result values aren't available while tracing!");
+        }
+        Result UseZero() override;
+        Result UseOne() override;
+
+        // -------------------------------------------------------------------------------------------------------------
+        // Instead of implementing IQuantumGateSet, the tracer provides 'tracing-by-id' methods. The QIR generation
+        // should translate all intrinsics to invoke these methods.
+        // The tracer doesn't differentiate between control and target qubits. However, While it could provide a single
+        // generic tracing method for and array of qubits, that would require the clients to copy control and target
+        // qubits into the same array. To avoid the copy, the tracer provides a method that takes two groups of qubits,
+        // where the first one can be empty or can be viewed as the set of controls.
+        // -------------------------------------------------------------------------------------------------------------
+        LayerId TraceSingleQubitOp(OpId id, Duration duration, Qubit target);
+        LayerId TraceMultiQubitOp(
+            OpId id,
+            Duration duration,
+            long nFirstGroup,
+            Qubit* firstGroup,
+            long nSecondGroup,
+            Qubit* secondGroup);
+
+        Result TraceSingleQubitMeasurement(OpId id, Duration duration, Qubit target);
+        Result TraceMultiQubitMeasurement(OpId id, Duration duration, long nTargets, Qubit* targets);
+        LayerId GetLayerIdOfSourceMeasurement(Result r) const
+        {
+            return reinterpret_cast<LayerId>(r);
+        }
+
+        // -------------------------------------------------------------------------------------------------------------
+        // Backing of the rest of the bridge methods.
+        // -------------------------------------------------------------------------------------------------------------
+        LayerId InjectGlobalBarrier(OpId id, Duration duration);
+
+        // -------------------------------------------------------------------------------------------------------------
+        // Configuring the tracer and getting data back from it.
+        // -------------------------------------------------------------------------------------------------------------
+        // Temporary method for initial testing
+        // TODO: replace with a safer accessor
+        const std::vector<Layer>& UseLayers()
+        {
+            return this->metricsByLayer;
+        }
+
+        void PrintLayerMetrics(std::ostream& out, const std::string& separator, bool printZeroMetrics) const;
+    };

```

Comment:
(At least in the future, feel free to ignore in this PR) 
* I would recommend to start separating the declaration and definition of the member functions, see example below. 
The code base is growing, and later it will be not very convenient to quickly get an idea of the class if the definition and declaration of the member functions is in the same place. 
* Also having modern (multi-pass) compilers it is typical to declare `public`s first, then `protected`s, then `private`s (even if this results in _use_ (of private member vars declared later, in a public member func defined earlier) _before declaration_).
* I would also recommend to keep the class definition _tight_, such that the whole class definition fits in one screen page (if it doesn't fit then the author may want to consider breaking the class up into smaller classes).
* It is typical to declare the (private) member vars after the private member funcs.

Example.
.hpp:
```c++
class CTracer : public ISimulator
{
  public:
    explicit CTracer(int preferredLayerDuration);
    CTracer(int preferredLayerDuration, const std::unordered_map<OpId, std::string>& opNames);

    // ISimulator:
    Qubit AllocateQubit() override;
    void ReleaseQubit(Qubit qubit) override;
    std::string QubitToString(Qubit qubit) override;
    void ReleaseResult(Result result) override;

    IQuantumGateSet* AsQuantumGateSet() override;
    IDiagnostics* AsDiagnostics() override;
    Result M(Qubit target) override;
    Result Measure(long numBases, PauliId bases[], long numTargets, Qubit targets[]) override;
    bool AreEqualResults(Result r1, Result r2) override;
    ResultValue GetResultValue(Result result) override;
    Result UseZero() override;
    Result UseOne() override;
    
    LayerId TraceSingleQubitOp(OpId id, Duration duration, Qubit target);   // -------------------------------------------------------------------------------------------------------------
    LayerId TraceMultiQubitOp(                                              // Instead of implementing IQuantumGateSet, the tracer provides 'tracing-by-id' methods. The QIR generation     
        OpId id,                                                            // should translate all intrinsics to invoke these methods.                                                     
        Duration duration,                                                  // The tracer doesn't differentiate between control and target qubits. However, While it could provide a single 
        long nFirstGroup,                                                   // generic tracing method for and array of qubits, that would require the clients to copy control and target    
        Qubit* firstGroup,                                                  // qubits into the same array. To avoid the copy, the tracer provides a method that takes two groups of qubits, 
        long nSecondGroup,                                                  // where the first one can be empty or can be viewed as the set of controls.                                    
        Qubit* secondGroup);                                                // -------------------------------------------------------------------------------------------------------------

    Result TraceSingleQubitMeasurement(OpId id, Duration duration, Qubit target);
    Result TraceMultiQubitMeasurement(OpId id, Duration duration, long nTargets, Qubit* targets);
    LayerId GetLayerIdOfSourceMeasurement(Result r) const;
    LayerId InjectGlobalBarrier(OpId id, Duration duration);    // Backing of the rest of the bridge methods.
    const std::vector<Layer>& UseLayers();  // Configuring the tracer and getting data back from it.
                                            // Temporary method for initial testing
                                            // TODO: replace with a safer accessor
    void PrintLayerMetrics(std::ostream& out, const std::string& separator, bool printZeroMetrics) const;

private:
    QubitState& UseQubit(Qubit q);
    const QubitState& UseQubit(Qubit q) const;
    LayerId FindLayerToInsertOperationInto(Qubit q, Duration opDuration) const;     // If no appropriate layer found, return `INVALID`
    LayerId CreateNewLayer(Duration minRequiredDuration);   // Returns the index of the created layer.
    void AddOperationToLayer(OpId id, LayerId layer);       // Adds operation with given id into the given layer. Assumes that duration contraints have been satisfied.
    void UpdateQubitState(Qubit q, LayerId layer, Duration opDuration); // Update the qubit state with the new layer information

private:
    std::vector<QubitState> qubits;     // For now the tracer assumes no reuse of qubits.
    const int preferredLayerDuration = 0;   // The preferred duration of a layer. An operation with longer duration will make the containing layer longer.
    std::vector<Layer> metricsByLayer;      // The index into the vector is treated as implicit id of the layer.
    LayerId globalBarrier = INVALID;        // The last global barrier, injected by the user. No new operations can be added to the barrier or to any of the
                                            // layer that preceeded it, even if the new operations involve completely new qubits.                           
    std::unordered_map<OpId, std::string> opNames;  // Mapping of operation ids to user-chosen names, for operations that user didn't name, the output will use
                                            // operation ids.
    std::unordered_set<OpId> seenOps;       // Operations we've seen so far (to be able to trim output to include only those that were encounted).
};
```

Reply:
My personal preference is to have all data members (private and public) to be declared at the top of the class. I don't know what's typical here because I worked on large codebases that had different takes as far as the order of declarations is concerned.

The readme.md in the root of QirRuntime attempts to capture the style guidelines and, while I did collect feedback from a few other people, it's biased towards my preferences. It's probably a good time now to review the guidelines, codify as many of them as possible into clang format and ensure consistency across the project.